T/121: Changelog for multiple packages in single repository

pomek opened the pull request:
### Merge message:

```
Feature: Introduced task for generating the changelog for multiple packages inside the single repository and fixed various issues related to generating the changelog. Closes #148. Closes #121. Closes #110. Closes #96.

BREAKING CHANGES: Function `executeOnDependencies()` is no longer available. Introduced `getSubPackagesPaths()` or `getSubRepositoriesPaths()` for gathering the package paths and `executeOnPackages()` for executing custom function on these packages.

BREAKING CHANGES: Task `tasks.generateChangelog()` has been renamed to `tasks.generateChangelogForSinglePackage()`.

BREAKING CHANGES: Task `generateChangelogForDependencies()` has been renamed to `tasks.generateChangelogForSubRepositories()`.

BREAKING CHANGES: Task `tasks.createRelease()` has been renamed to `tasks.createReleaseForSubRepository()`.

BREAKING CHANGES: Task `tasks.releaseDependencies()` has been renamed to `tasks.releaseSubRepositories()`.

NOTE: Introduced new task - `tasks.generateChangelogForSubRepositories()`.
```

Issues mentioned in the merge message: #148, #121, #110, #96.

---

In order to test this tool, I suggest cloning my repository – https://github.com/pomek/ckeditor5-dev.
Please remember you need to create a release commits (and tags) manually. 

This PR is a missing part of #144. 

---

First call this tool in `ckeditor5-dev` package will take around 10 min because the whole repository will be parsed. If you want to avoid it, you need to create a CHANGELOG file for each package with an entry for the latest version.

pomek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js:
```diff
@@ -5,66 +5,98 @@
+		for ( const item of commits ) {
+			const singleCommit = transformCommit( item, context );
 
-		conventionalRecommendedBump( options, parserOpts, ( err, response ) => {
```
Because `conventionalRecommendedBump` does not allow passing custom options to `git log` command, I had to remove this package and rewrite it. 

Thanks to that I could resolve #96.

pomek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/hascommitsfromlastrelease.js:
```diff
@@ -1,30 +0,0 @@
- *
- * @returns {Boolean}
- */
-module.exports = function hasCommitsFromLastRelease() {
```
It is not needed because [`getNewReleaseType()`]()https://github.com/ckeditor/ckeditor5-dev/pull/154/files#diff-81ddd240a176823f737da5a270374c71 knows all commits and can decide about the release type.

pomek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5devpackage.js:
```diff
@@ -0,0 +1,45 @@
+ * @param {Object} context.packageData Content from the 'package.json' for given package.
+ * @returns {Commit}
+ */
+module.exports = function transformCommitForCKEditor5DevPackage( commit, context ) {
```
We accept the commit only if its files are located in current processing package. [See the tests.](https://github.com/ckeditor/ckeditor5-dev/pull/154/files#diff-c4d9754e1f61e34742144bb2dcc0d345R62)

Reinmar replied:
Isn't there more gitty way of retrieveing such commits? Like in http://stackoverflow.com/questions/16343659/how-to-show-git-log-history-for-a-sub-directory-of-a-git-repo?

pomek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/versions.js:
```diff
@@ -17,11 +17,19 @@ const versions = {
+			return null;
+		}
+
+		return matches[ 2 ] ? matches[ 2 ] : matches[ 1 ];
```
Now we handle both types of changelog entry:

1. For entry generated for the first time:
```md
CHANGELOG
==========

## version (date)
```

2. For the next entries:
```md
CHANGELOG
==========

## [version](URL to compare the changes) (date)
```

pomek replied:
It fixes #148.

Reinmar replied:
This condition should be unnecessary. It wasn't necessary when I patched this in https://github.com/ckeditor/ckeditor5-dev/compare/t/148?expand=1.

pomek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/executeondependencies.js:
```diff
@@ -26,7 +28,7 @@ const getPackageJson = require( './getpackagejson' );
 module.exports = function executeOnDependencies( options, functionToExecute ) {
 	const packagesAbsolutePath = path.join( options.cwd, options.packages );
-	const directories = workspaceUtils.getDirectories( packagesAbsolutePath );
+	const directories = tools.getDirectories( packagesAbsolutePath );
```
Because we check whether the directory is a valid package, we can take everything from the CWD (instead of directories starting with `ckeditor5-`).

pomek commented:
I left a few comments which should explain why I did something this way.

ma2ciek commented:
I'm going to make a big, big coffee :D

ma2ciek commented on packages/ckeditor5-dev-env/lib/index.js:
```diff
@@ -26,26 +26,57 @@ const tasks = {
+			.then( () => {
+				const packageNames = Object.keys( generatedChangelog );
+
+				if ( !packageNames.length ) {
```
Don't we want to log here something as well?

pomek replied:
Not sure if we need an information when we skipped all packages.

Reinmar replied:
Let's leave it – it's very unlikely anyway.

ma2ciek commented on packages/ckeditor5-dev-env/lib/index.js:
```diff
@@ -26,26 +26,57 @@ const tasks = {
 			cwd: options.cwd,
 			packages: options.packages,
-			skipPackages: options.skipPackages || []
+			skipPackages: options.skipPackages || [],
```
`skippedPackages`

pomek replied:
For what? "Skip packages" is a list of packages to skip.  "Skipped packages" means "list of packages that have been skipped".

pomek replied:
Instead of `skipPackages`, it could be `packagesToSkip`. But it is not a part of this PR.

ma2ciek replied:
AFAIK skip can be used as a verb / noun. But it's rather just a naming convention. `skipPackages` suggests that the variable is a function.

Reinmar replied:
I think that since this is a config option and not a property of some stateful object, it's fine that it's called like the `--skip` option. Besides, it may even become a `{Function|Array}` one day and what's then?

ma2ciek commented on packages/ckeditor5-dev-env/lib/release-tools/tasks/generatechangelog.js:
```diff
@@ -91,12 +111,29 @@ module.exports = function generateChangelog( newVersion = null ) {
 		}
 	} );
+
+	function getTransformFunction() {
```
this function may be easily moved outside the main function

ma2ciek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/transform-commit-utils.js:
```diff
@@ -0,0 +1,96 @@
+	 * @param {String} value
+	 * @returns {String}
+	 */
+	linkGithubUsers( value ) {
```
A parameter could be more precise.

ma2ciek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5devpackage.js:
```diff
@@ -0,0 +1,45 @@
+ * Parses a single commit for `ckeditor5-dev` package.
+ *
+ * @param {Commit} commit
+ * @param {Object} {context}
```
small typo - `@param {Object} context`

ma2ciek commented on packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js:
```diff
@@ -0,0 +1,144 @@
+ *   - makes links to issues and user's profiles on GitHub.
+ *
+ * @param {Commit} commit
+ * @param {Object} {context}
```
and here as well

ma2ciek commented:
For me whole PR looks very good. Would you have a look @Reinmar?

pomek commented:
This PR is required for me before I start resolving other tickets. 

I modified a lot of files and I moved a lot of files from their original localization. I would like to avoid conflicts between changed files when we close another ticket that changed already moved a file in this PR.

Reinmar commented:
I'll be reviewing it today but most likely it will take me until the end of the week. You're on holidays anyway, so it'll be fine, right?

pomek commented:
Yes. Will be nice to have it next week.

Reinmar commented:
The biggest issue with this PR is that it tries to squash everything into one function – `generateChangelogForDependencies()` and set parametrise it through options. This means that this single function becomes very complex and that these changes affect how it worked in the past. It's far harder now to validate these changes.

DRY doesn't mean that we must strive to create the least number of functions. It can actually mean the opposite if KISS is involved.

In this case, we have clear two entry points – generating changelog for sub repos and generating changelog for sub packages. While they may share a lot of the same logic (which can be extracted to some meaningful utils), they also differ significantly. You need to identify the parts which differ and implement them separately. Avoid config options – that should be your rule of thumb.

E.g if one of these functions need to verify something with `package.json`, then only this function need to call some `verifyIsDependency()` function. The other function shouldn't be concerned by it.

Another thing is actually finding the right "generalisation" of te problem you're dealing with. You introduced option `isDevPackage` forcing somehow that this is only used for dev packages. If, we'd have another repo which is similar to ckeditor5-dev, we shouldn't use that option... so what's then?

Hence, you need to do a step back and look at the situation. First of all, the current name `generateChangelogForDependencies()` already imposes something – it's about dependencies. Then you have `executeOnDependencies()` which should also be about dependencies. But are they in the case of ckeditor5-dev? Nope, they are not. You even needed to introduce the `checkPackageJson` option denying these functions' names.

So, what's a better abstraction? There are many options, but my first choice would be:

* `changelog.generateForSubRepositories()` and `changelog.generateForSubPackages()`
* `executeOnPackages()` and `getSubRepositories()` and `getSubPackages()`. Note that the difference between fetching sub repos and sub packages will be handled by the two last functions. The first function's job should be to run code on them which is in line with SRP and which actually may make it unnecessary.

Going further, we can see that we will have functions which are generic (like `executeOnPackages()`) and functions which will server one of the two modes – sub repos and sub packages. Depending on number of them it may be reasonable to e.g. split the `utils/*` to `utils/*`, `utils/packages/`, `utils/repositories/`. It's best to do such things once situation clarifies.

To sum up – I'll perform a small review, but the first thing to do is to clean up the code anyway. Only then it will be possible to check everything, so right now I'll just focus on more visible things.

Reinmar commented on packages/ckeditor5-dev-env/lib/index.js:
```diff
@@ -26,26 +26,57 @@ const tasks = {
 			packages: options.packages,
-			skipPackages: options.skipPackages || []
+			skipPackages: options.skipPackages || [],
+			checkPackageJson: options.checkPackageJson === undefined ? true : options.checkPackageJson
```
Optional boolean flags should be `false` by default to avoid such a code and just because it's more intuitive that when you pass some option you enable something. So, in this case it should be `skipPackageJsonCheck`. But of course, the whole option will rather be unnecessary after refactor.

Reinmar commented on packages/ckeditor5-dev-env/lib/release-tools/tasks/generatechangelog.js:
```diff
@@ -91,12 +111,29 @@ module.exports = function generateChangelog( newVersion = null ) {
+				let commitMessage;
+
+				if ( options.isDevPackage ) {
+					commitMessage = `Changelog for "${ packageJson.name }".`;
```
Why committing every changelog separately? That's exactly one releasing sub packages should work differently than sub repos – in that case, there should be just one commit. Taken the number of code branching in this function you may have a perfect proof that it just shouldn't be the same function.

Reinmar commented on packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js:
```diff
@@ -5,66 +5,98 @@
+
+	if ( fromVersion ) {
+		if ( isDevPackage ) {
+			fromVersion = packageJson.name + '@' + fromVersion;
```
I'm not sure about the fact that in one case we use `vx.y.z` notation and in the other case there's no `v`. I see that Lerna doesn't use `v` in its tags... so we can follow that, but then let's this logic extracted somewhere because I can see that the same code is repeated twice.

pomek commented:
Guys (@Reinmar, @ma2ciek), ready to review. I will modify the proposed merge message.

Reinmar commented:
I merged the code and... it turned out that the new task doesn't work. Running `gulp changelog` in ckeditor5-dev does nothing now. It worked on t/121 though. The task may have problems with reading merge commits.

Reinmar commented:
Hm... I don't know why it works, but the commit is apparently rejected at some point. `gitRawCommits()` return the proper stuff:

```
(master fefc1de ?1) p@m /workspace/ckeditor5-dev> git log --format="%B%n-hash-%n%H%n------------------------ >8 ------------------------" @ckeditor/ckeditor5-dev-env@4.4.3..HEAD --first-parent
Merge pull request #154 from ckeditor/t/121

Feature: A task for generating changelogs in a monorepo was introduced. Several other improvements were made on the occasion. Closes #148. Closes #121. Closes #110. Closes #96.

BREAKING CHANGES: Task `tasks.generateChangelog()` has been renamed to `tasks.generateChangelogForSinglePackage()`.

BREAKING CHANGES: Task `generateChangelogForDependencies()` has been renamed to `tasks.generateChangelogForSubRepositories()`.

BREAKING CHANGES: Task `tasks.createRelease()` has been renamed to `tasks.releaseRepository()`.

BREAKING CHANGES: Task `tasks.releaseDependencies()` has been renamed to `tasks.releaseSubRepositories()`.

NOTE: Introduced a new task – `tasks.generateChangelogForSubRepositories()`.
-hash-
fefc1de92d89761e30cbc1b23d09ece0ce6403f8
------------------------ >8 ------------------------
```

So, the problem lays somewhere later. Most likely, in getting the list of changed files:

```
# normal commit:
(master fefc1de ?1) p@m /workspace/ckeditor5-dev> git show --pretty="" --name-only 118fccb2
packages/ckeditor5-dev-env/lib/release-tools/utils/versions.js
packages/ckeditor5-dev-env/tests/release-tools/utils/changelog.js
packages/ckeditor5-dev-env/tests/release-tools/utils/versions.js
# merge commit:
(master fefc1de ?1) p@m /workspace/ckeditor5-dev> git show --pretty="" --name-only fefc1de
```
